### ISSUE TITLE
Incorrect segment counting

### ISSUE THREAD

**Issue by kasyrm** [TwilioDevEd/message-segment-calculator#39]
I spotted a similar issue to this one: https://github.com/TwilioDevEd/message-segment-calculator/issues/18.

When providing a message like this one: 
`]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]`

Segment calculator shows it's 3 segments. 

It's 153 scalars counted as 306 characters. That's 2142 bits for characters + 96 bits for 2 headers. That should add up to 2238 bits so still within 2 segments.

But when adding the last character of this string another header is added and the message takes up 2286 bits: 3 segments. 

There are other calculators that show this as 2 segments (https://sakari.io/sms-length-calculator/)

But when sending the message on Twillio it is sent as 3 segments. So I have been wondering which is correct and what is causing this behaviour.

**Comment by jarangutan** [comment]
I don't believe that the segment calculator is incorrect. For GSM-7 encoding, `]` is hex `1B3E` which is reflected accurately as `0x001B 0x003E`. See here https://www.developershome.com/sms/gsmAlphabet.asp

**Comment by amanovofficial** [comment]
> ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]

Checked out this string using https://messente.com/documentation/tools/sms-length-calculator
it shows that the string splits into 2 segments but your calculator considers the string as 3 segments.

**Comment by StandBackBurrito** [comment]
> > ]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
> 
> Checked out this string using https://messente.com/documentation/tools/sms-length-calculator it shows that the string splits into 2 segments but your calculator considers it as 3 segments.

This site says 2 parts, but it has 3 headers with a single `]` after the last header. 

![image](https://github.com/user-attachments/assets/5ebb48e7-1c72-496e-8f21-6734506f7616)

**Comment by thinkingserious** [comment]
After investigating the codebase, the calculator is **correct** here, and this behavior matches how Twilio actually sends the message (as you confirmed — Twilio sends it as 3 segments).

**Why it's 3 segments, not 2:**

Each `]` is a GSM-7 extension character requiring 14 bits (escape + char). With 153 `]` characters:
- Theoretical bits needed: 153 × 14 = 2,142 bits
- Available in 2 segments with UDH: 2 × (1,120 − 48) = 2,144 bits

So it *theoretically* fits in 2 segments. However, segment boundaries must fall between characters, not mid-character. After adding the 48-bit UDH to segment 1, only 1,072 bits remain — enough for 76 characters (76 × 14 = 1,064), leaving 8 unused bits that can't fit another 14-bit character. Each segment holds at most 76 `]` characters, so 153 characters requires 3 segments.

Other calculators showing 2 segments are incorrect. Closing as working-as-intended.
